File: backend/app/models/transcription.py
from pydantic import BaseModel, ConfigDict


class TranscribeRequest(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    lang_code: str
    model_card: str = "omniASR_CTC_1B_v2"
    sample_idx: int = 0
    audio_base64: str | None = None


class TranscribeResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    lang_code: str
    model_card: str
    prediction: str
    ground_truth: str
    cer: float
    latency_ms: int


class CompareRequest(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    lang_code: str
    sample_idx: int = 0
    model_cards: list[str] = [
        "omniASR_CTC_300M_v2",
        "omniASR_CTC_1B_v2",
        "omniASR_LLM_1B_v2",
        "omniASR_LLM_7B_v2",
    ]


class CompareResponse(BaseModel):
    lang_code: str
    results: list[TranscribeResponse]

File: frontend/src/api/transcription.ts
import client from "./client";

export interface TranscribeResult {
  lang_code: string;
  model_card: string;
  prediction: string;
  ground_truth: string;
  cer: number;
  latency_ms: number;
}

export interface CompareResult {
  lang_code: string;
  results: TranscribeResult[];
}

export async function transcribe(
  langCode: string,
  modelCard: string = "omniASR_CTC_1B_v2",
  sampleIdx: number = 0
): Promise<TranscribeResult> {
  const { data } = await client.post<TranscribeResult>("/transcribe", {
    lang_code: langCode,
    model_card: modelCard,
    sample_idx: sampleIdx,
  });
  return data;
}

export async function compareModels(
  langCode: string,
  sampleIdx: number = 0,
  modelCards?: string[]
): Promise<CompareResult> {
  const { data } = await client.post<CompareResult>("/transcribe/compare", {
    lang_code: langCode,
    sample_idx: sampleIdx,
    ...(modelCards ? { model_cards: modelCards } : {}),
  });
  return data;
}

export function getAudioUrl(langCode: string, sampleIdx: number = 0): string {
  return `/api/v1/transcribe/audio/${langCode}?sample_idx=${sampleIdx}`;
}
